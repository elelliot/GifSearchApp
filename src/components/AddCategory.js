import React, { useState } from 'react';
import PropTypes from 'prop-types';

export const AddCategory =  ( {setCategories} ) => {

    const [inputValue, setInputValue] = useState('');//string vacío, si lo dejamos sin '', es undefined y causará errores

    const handleInputChange = ( e ) => {
        //console.log(e.target.value);
        setInputValue (e.target.value);//obtenemos los cambios de texto con el evento

        //console.log('handleInputChange llamado');
    }

    const handleSubmit = ( e ) =>{

        e.preventDefault()//prevenimos el comportamiento por defecto que sería refrescar
        // console.log('Submit Hecho')

        //console.log('handleSubmit', inputValue);


        //validamos que no haya valores nulos o que sean mínimo de 3 caracteres
        if( inputValue.trim().length > 2){
            
            /*como no pasé las categories como argumento junto con el setCategories 
            llamo al callback que tiene el estado anterior, en este caso le llamo cats*/
            setCategories( cats => [ inputValue, ...cats]);//insertamos primero la categoría que buscamos y despues las anteriores
            setInputValue('');
        }


    }


    return (
        //cuando se quiere regresar un formulario, los fragments son redundantes ya que el form agrupa o el div
        //cuando aplastas enter el form se envía y se refresca por defecto
        <form onSubmit = { handleSubmit }>
            <p> { inputValue }</p>
            <input 
                type="text"
            /*el value se queda igual ya que esto se redibuja cada que toco el teclado y agarramos el valor desde el evento onChange*/
               value= { inputValue }
            /*Se puede mandar el primer argumento del onChange que sería el evento y lo atrapamos en la función (arriba) */
                onChange = { handleInputChange /*e => console.log(e)*/ } 
            />
        </form>
        
    )
}
//hacemos que el setCategories sea funcion obligatoria
AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}
