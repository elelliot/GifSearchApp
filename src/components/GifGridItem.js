import React from 'react'
import PropTypes from 'prop-types';

export const GifGridItem = ( { /*id,*/ title, url } ) => {

    // console.log( id, title, url  );

    return (
        <div className="card animate__animated animate__fadeIn ">
            <img src= { url } alt = { title } />
            <p>{ title }</p>
        </div>
    )
}

GifGridItem.propTypes = {
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired,
}
// import React from 'react'

// export const GifGridItem = ( img ) => {

//     console.log( img );
    
//     return (
//         <div>
//             { img.title }
//         </div>
//     )
// }

















