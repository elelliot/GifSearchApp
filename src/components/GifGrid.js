import React from 'react';
import Proptypes from 'prop-types';

import { useFetchGifs } from '../hooks/useFetchGifs'
// import { getGifs } from '../helpers/getGifs';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({ category }) => {

    //usamos el custom hook
    const { data:images , loading } = useFetchGifs( category ); //renombramos la data desestructurada


    // const [images, setImages] = useState([])


    return (
        <>
            <h3 className=" animate__animated animate__fadeIn ">{ category }</h3>

            {/* para no tener que poner un null si la condicion no se cumple usamos el && para solo evaluar lo primero 
            y no hacer nada si no se cumple */}
            { loading && <p className = "animate__animated animate__flash">Loading</p> }
            {/*  loading ? <p>Loading</p> : null */}
            
            <div className = "card-grid">
                
                {
                    images.map( img => (
                        <GifGridItem 
                            key = { img.id }
                            { ...img } 
                            // img = { img } 
                        />
                        // return <li key= { imagen.id }>{ imagen.title }</li>
                    ))
                    
                }
                
            </div>
        </>
    )
}


GifGrid.propTypes = {
    category: Proptypes.string.isRequired
}

