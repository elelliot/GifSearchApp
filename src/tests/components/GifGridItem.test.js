import { shallow } from "enzyme"
import { GifGridItem } from "../../components/GifGridItem"



describe('Pruebas en <GifGidItem />', () => {

    const title = 'Un titulo';
    const url = 'https://localhost/algo.jpg';
    
    const wrapper = shallow( <GifGridItem title = { title } url={ url }/>)

    test('Debe mostrar el componente correctamente ', () => {
        

        expect( wrapper ).toMatchSnapshot();
    })
    


    test('Debe de tener un parrafo con el title', () => {
        

        const p = wrapper.find('p');
        expect( p.text().trim() ).toBe( title );

    });


    test('Debe tener la imagen igual al url y alt de los props', () => {
        

        const img = wrapper.find('img'); //como solo es una imagen se puede hacer eso, si no, buscarla por id o clase
        
        //1-  img.html() --- te devuelve la etiqueta con las propiedades 
        //2-  img.props() --- te devuelve un objeto con las props
        //2.5 img.props().src --- te devuelve la propiedad del tag (src en este caso)
        //3-  img.prop('src') --- lo mismo pero solo te devuelve una propiedad
        
        expect( img.prop('src') ).toBe( url );
        expect( img.prop('alt') ).toBe( title );
    });
    

    test('Debe tener animate__fadeIn', () => {
        
        const div = wrapper.find('div');

        //.hasClass devuelve true si contiene al menos 1 clase y el toBeTrurthy confirma que sea true por defecto en lugar de ponerle toBe( true )
        expect( div.hasClass('animate__fadeIn') ).toBeTruthy();
    })
    
    
})
