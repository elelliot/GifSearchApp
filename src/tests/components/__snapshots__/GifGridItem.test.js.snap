// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifGidItem /> Debe mostrar el componente correctamente  1`] = `
<div
  className="card animate__animated animate__fadeIn "
>
  <img
    alt="Un titulo"
    src="https://localhost/algo.jpg"
  />
  <p>
    Un titulo
  </p>
</div>
`;
